pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -839,8 +825,37 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
 						Service:                    service,
 					}
 
-					configgen.buildSidecarOutboundListenerForPortOrUDS(node, listenerOpts, pluginParams, listenerMap,
-						virtualServices, actualWildcard)
+					// Minor optimization for Kubernetes stateful sets/headless services
+					// Instead of generating a single 0.0.0.0:Port listener, generate N listeners
+					// (where N <= 5 at most) one for each podIP.
+					if bind == "" && service.Resolution == model.Passthrough &&
+						service.Attributes.ServiceRegistry == string(serviceregistry.KubernetesRegistry) {
+						instances, err := env.InstancesByPort(service, servicePort.Port, nil)
+						if err != nil || len(instances) == 0 || len(instances) > 5 {
+							// we can't do anything. Fallback to the usual way of constructing listeners
+							// for headless services that use 0.0.0.0:Port listener
+							configgen.buildSidecarOutboundListenerForPortOrUDS(node, listenerOpts, pluginParams, listenerMap,
+								virtualServices, actualWildcard)
+						} else {
+							for _, instance := range instances {
+								listenerOpts := buildListenerOpts{
+									env:            env,
+									proxy:          node,
+									proxyInstances: node.ServiceInstances,
+									proxyLabels:    proxyLabels,
+									port:           servicePort.Port,
```

Comment:
change to instance.Endpoint.Port ? Think of that when the service port not equals to targetPort.

Reply:
err no. this is for outbound, where we use service ports only

Reply:
But we use this for listener address

```
	return &xdsapi.Listener{

		Address:         util.BuildAddress(opts.bind, uint32(opts.port)),

	}
```